libgalois/include/katana/PropertyGraph.h

```diff
@@ -97,6 +97,29 @@ struct KATANA_EXPORT GraphTopology {
 /// manages the serialization of the various partitions and properties that
 /// comprise the physical representation of the logical property graph.
 class KATANA_EXPORT PropertyGraph {
+public:
+  /// TypeID uniquely identifies/contains a combination/set of types
+  /// TypeID is represented using 8 bits
+  /// TypeID for nodes is distinct from TypeID for edges
+  using TypeID = uint8_t;
```

Comment:
Maybe but I didn't think it was necessary, so didn't attempt that in this PR.

Reply:
That's interesting. TypeID represents/identifies a collection of type names. We can think of better names. Any suggestions?

Reply:
Unions of types are types. So we could call this `UnionTypeID`. Another option that would make sense to non-type theorists would be `TypeSetID`.

Reply:
I like `TypeSetID` more but we have a set of current `TypeID`s called `TypeIDsSet`. If we go with that, it would be `TypeSetIDsSet`. Is that confusing?

Reply:
Ok, makes sense.

Reply:
I vote against TypeSetIDSet. It's a mouthful and I find it confusing. We have 2 things at hand:
1. Type IDs 
2. Sets of Type IDs (call them TypeIDSet or what you prefer), which are represented as TypeIDs themselves (just a different integer for each set of type IDs in the space of possible type IDs). But we can also represent these sets as bit sets.

Reply:
I don't mind either name but I tend toward `TypeID` and `TypeIDsSet` because even though `TypeID` refers to a set of types, in practice it refers to a single derived class-type (the other classes are base class-types). We can revisit this name later because I don't think we have a consensus regarding the terms: type, tag, and label.

Reply:
I'm pretty strongly against using `TypeID` to represent a set of `TypeID`s. I'm starting to wonder if I misunderstand what's going on here. So let me explain what I think is going on, and then people can tell me if I missed the point and maybe the text will be useful for documentation.

There are 3 things we are dealing with here and trying to create types for:

1. Actual types (also known as labels) which are usually represented as a boolean indicator property in the RDG. We don't really directly handle these in the code except during loading. Let's call these atomic types.
2. Sets of atomic types which are found by looking at the atomic types that entities have in the data set and finding sets that appear together (e.g., Doctor and Person might be a group because every Doctor is also a Person). These are used in our compressed property representation (which isn't implemented yet, IIUC, but will come along eventually). They are not used strictly as intersections (I think I said union above when I meant intersection, BTW) of their constituent types, IIRC, because entities are associated with a set of atomic types may actually have most but not all of the types in the set.
3. Sets of sets of atomic types. I don't know what these are used for, but I thought Roshan said we had these.

My understanding was that `TypeID` (which I pushed to call `TypeSetID`) was a numeric ID for 2 above (a set of atomic types). And that `TypeSetIDSet` was an actual collection (`std::vector` or whatever) of `TypeSetID`s and hense would be 3 above.

I'm starting to really doubt myself. So please tell me what the reality here is.

Reply:
More detailed response after the meeting but your understanding of 2 and 3 is correct.

Reply:
@arthurp : Your understanding is correct. I'll add this (and you may already have understood this) that 1 and 2 above are represented as integers in the same 8 bit space. I agree with the idea of 3 and it's being used in query code (the other PR), but not sure if we need to define such a type in PropertyGraph.

Reply:
I didn't realize that we assigned numeric ID to both atomic types and the sets of types. I don't actually like that because from my understanding the sets of types are not always true intersection types since an entity may be assigned to a set even if it doesn't actually have every type in the set. If I'm wrong about that and entities are only assigned to a set of types if they have all the types in the set, then they are intersection types and I'm fine with having a single ID space for all types as long as we make sure we are aware of the interactions between atomic and intersection types.

Reply:
@amberhassaan Two corrections:
(A) We do NOT assign numeric ID to (1). A type is recognized by its name (just like a property). 
(B) We need to define (3) in PropertyGraph because it maintains a map from a type-name to (3). 

@arthurp One correction:
(C) I'm not sure if union/intersection is an appropriate term to use here but the way you used it seemed like (2) always corresponds true intersection types. A node has a numeric ID corresponding to (2) if and only if the set of all its atomic types is strictly equal to the set identified by that numeric ID.

Reply:
Listening to @arthurp's comments, I'm now more inclined towards `TypeSetID` and `TypeSetIDSet` for (2) and (3) respectively.

Reply:
@roshandathathri : aren't you assigning 8 bit IDs to individual types as well?

Reply:
After looking at the code, here are the names I'm going for:
```cpp
  /// TypeSetID uniquely identifies/contains a combination/set of types
  /// TypeSetID is represented using 8 bits
  /// TypeSetID for nodes is distinct from TypeSetID for edges
  using TypeSetID = uint8_t;
  using ArrowTypeSetID = arrow::UInt8Type;
  static constexpr TypeSetID kUnknownType = TypeSetID{0};
  static constexpr TypeSetID kInvalidType =
      std::numeric_limits<TypeSetID>::max();
  /// A set of TypeSetIDs
  using SetOfTypeSetIDs =
      std::bitset<std::numeric_limits<TypeSetID>::max() + 1>;
  /// A set of type names
  using SetOfTypeNames = std::unordered_set<std::string>;
  /// A map from TypeSetID to the set of the type names it contains
  using TypeSetIDToSetOfTypeNamesMap =
      std::vector<SetOfTypeNames>;
  /// A map from the type name to the set of the TypeSetIDs that contain it
  using TypeNameToSetOfTypeSetIDsMap =
      std::unordered_map<std::string, SetOfTypeSetIDs>;
```
It's verbose but more clear IMO.

Reply:
@amberhassaan No, that's (2) - set of types.

Reply:
>  I'm not sure if union/intersection is an appropriate term to use here but the way you used it seemed like (2) always corresponds true intersection types. A node has a numeric ID corresponding to (2) if and only if the set of all its atomic types is strictly equal to the set identified by that numeric ID.

Yep, those are intersection types. A specific kind of intersection type, but still intersection types. https://en.wikipedia.org/wiki/Intersection_type